src/simulator/simulator.rs

```diff
@@ -342,16 +344,11 @@ impl FluidSim {
     /// calculates the indexes and returns the in the top, right, bottom, left order
     fn indexes_around(&self, x_index: usize, y_index: usize) -> [usize; 4] {
         [
-            (x_index, y_index + 1),
-            (x_index + 1, y_index),
-            (x_index, y_index - 1),
-            (x_index - 1, y_index),
+            self.calculate_index(x_index, y_index + 1),
+            self.calculate_index(x_index + 1, y_index),
+            self.calculate_index(x_index, y_index - 1),
+            self.calculate_index(x_index - 1, y_index),
         ]
-        .into_iter()
-        .map(|(x, y)| self.calculate_index(x, y))
-        .collect::<Vec<usize>>()
-        .try_into()
-        .unwrap()
     }
```

Comment:
This is where the majority of improvement comes from.
Rust was unable to optimize out the allocation of this vector, and since this function was called for every single cell, this is the hot path.

Reply:
I don't know how you found this, but this is a great improvement!